sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/AbstractInstrumentBuilder.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics;
+
+import io.opentelemetry.api.metrics.ObservableDoubleMeasurement;
+import io.opentelemetry.api.metrics.ObservableLongMeasurement;
+import io.opentelemetry.sdk.metrics.common.InstrumentDescriptor;
+import io.opentelemetry.sdk.metrics.common.InstrumentType;
+import io.opentelemetry.sdk.metrics.common.InstrumentValueType;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+
+/** Helper to make implementing builders easier. */
+public abstract class AbstractInstrumentBuilder<BuilderT extends AbstractInstrumentBuilder<?>> {
+  private final MeterProviderSharedState meterProviderSharedState;
+  private final MeterSharedState meterSharedState;
+  private final String instrumentName;
+  private String description;
+  private String unit;
+
+  AbstractInstrumentBuilder(
+      MeterProviderSharedState meterProviderSharedState,
+      MeterSharedState meterSharedState,
+      String name,
+      String description,
+      String unit) {
+    this.instrumentName = name;
+    this.description = description;
+    this.unit = unit;
+    this.meterProviderSharedState = meterProviderSharedState;
+    this.meterSharedState = meterSharedState;
+  }
+
+  protected abstract BuilderT getThis();
+
+  public BuilderT setUnit(String unit) {
+    this.unit = unit;
+    return getThis();
+  }
+
+  public BuilderT setDescription(String description) {
+    this.description = description;
+    return getThis();
+  }
+
+  @FunctionalInterface
+  protected static interface SwapBuilder<T> {
```

Comment:
this interface is hanging out in the middle of the class. I think our standards say to put it at the bottom.

Reply:
moved.  Debated adding ascii-art-lawn-chair to denote its chill attitude towards live (and high coupling to the constructors required for builders).